devdoc/interlocked_requirements.md

```diff
@@ -0,0 +1,298 @@
+MOCKABLE_FUNCTION(, void*, interlocked_exchange_pointer, volatile void*, target, void*, value);
+MOCKABLE_FUNCTION(, int32_t, interlocked_increment, volatile int32_t*, addend);
+MOCKABLE_FUNCTION(, int16_t, interlocked_increment_16, volatile in       t16_t*, addend);
+MOCKABLE_FUNCTION(, int64_t, interlocked_increment_64, volatile int64_t*, addend);
+MOCKABLE_FUNCTION(, int32_t, interlocked_or, volatile int32_t*, destination, int32_t, value);
+MOCKABLE_FUNCTION(, int16_t, interlocked_or_16, volatile int16_t*, destination, int16_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_or_64, volatile int64_t*, destination, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_or_8, volatile int8_t*, destination, int8_t, value);
+MOCKABLE_FUNCTION(, int32_t, interlocked_xor, volatile int32_t*, destination, int32_t, value);
+MOCKABLE_FUNCTION(, int16_t, interlocked_xor_16, volatile int16_t*, destination, int16_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_xor_64, volatile int64_t*, destination, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_xor_8, volatile int8_t*, destination, int8_t, value);
+```
+
+## interlocked_add
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_add, volatile int32_t*, addend, int32_t, value);
+```
+
+**SRS_INTERLOCKED_43_001 [** `interlocked_add` shall perform an atomic addition operation on the specified 32-bit integer values.**]**
+
+
+
+## interlocked_and
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_and, volatile int32_t*, destination, int32_t, value);
+```
+
+**SRS_INTERLOCKED_43_002 [** `interlocked_and` shall perform an atomic AND operation on the specified 32-bit integer values.**]**
+
+## interlocked_and_16
+
+```c
+MOCKABLE_FUNCTION(, int16_t, interlocked_and_16, volatile int16_t*, destination, int16_t, value);
+```
+
+**SRS_INTERLOCKED_43_003 [** `interlocked_and_16` shall perform an atomic AND operation on the specified 16-bit integer values.**]**
+
+
+## interlocked_and_64
+
+```c
+MOCKABLE_FUNCTION(, int64_t, interlocked_and_64, volatile int64_t*, destination, int64_t, value);
+```
+
+**SRS_INTERLOCKED_43_004 [** `interlocked_and_64` shall perform an atomic AND operation on the specified 64-bit integer values.**]**
+
+## interlocked_and_8
+
+```c
+MOCKABLE_FUNCTION(, int64_t, interlocked_and_8, volatile int8_t*, destination, int8_t, value);
+```
+
+**SRS_INTERLOCKED_43_005 [** `interlocked_and_8` shall perform an atomic AND operation on the specified 8-bit integer values.**]**
+
+## interlocked_compare_exchange
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_compare_exchange, volatile int32_t*, destination, int32_t, exchange, int32_t, comperand);
+```
+
+**SRS_INTERLOCKED_43_006 [** `interlocked_compare_exchange` shall perform an atomic compare-and-exchange operation on the specified values. The function compares two specified 32-bit values and exchanges with another 32-bit value based on the outcome of the comparison.**]**
+
+## interlocked_compare_exchange_128
```

Comment:
>interlocked_compare_exchange_128 [](start = 3, length = 32)

seems to only be available for 64 bit architectures, such as x64. ARM64.

what would happen if it is used in a 32 bit process? Does it fail the compilation?

Reply:
I'm not sure what will happen because MSDN doesn't say anything about it. Dan was saying that we only need to worry about 64 bit architectures because that's where this code will run


---
In reply to: [424109993](https://github.com/Azure/azure-c-util/pull/39#discussion_r424109993) [](ancestors = 424109993)

Reply:
We only support x64 for block store anyway (I know, this is a more generic library, but maybe it needs the same restriction anyway). I think this is fine as Dan said.

---
In reply to: [425466509](https://github.com/Azure/azure-c-util/pull/39#discussion_r425466509) [](ancestors = 425466509,424109993)